### ISSUE TITLE
Installing gdev.ko module fails.

### ISSUE THREAD

**Issue by itziksh** [shinpei0208/gdev#6]
Hi,
We are installing gdev on ubuntu 12.04 64 bit.
We compiled kernel 3.3.0 according to the instructions at https://gist.github.com/3057893 and booted from it.
We installed envytools and pscnv (from an older commit, d0bdb2f)  as well.

When trying to proceed with instructions to install Gdev (https://github.com/shinpei0208/gdev , step 4), the make succeeded with warnings, and we got an error when trying to load gdev.ko (insmod gdev.ko)-
insmod: error inserting 'gdev.ko': -1 Unknown symbol in module
This modules which might be needed are loaded-

drm
drm_kms_helper
i2c_algo_bit
video
pscnv.ko

envytools is not loaded as we black-listed it.

When making gdev.ko we received the following output-

make -C /lib/modules/3.3.0-custom+/build M=/homes/uriv/gdev/mod/build modules
make[1]: Entering directory /homes/uriv/linux-2.6'
CC [M] /homes/uriv/gdev/mod/build/gdev_drv.o
CC [M] /homes/uriv/gdev/mod/build/gdev_drv_nvidia.o
CC [M] /homes/uriv/gdev/mod/build/gdev_fops.o
CC [M] /homes/uriv/gdev/mod/build/gdev_ioctl.o
CC [M] /homes/uriv/gdev/mod/build/gdev_proc.o
CC [M] /homes/uriv/gdev/mod/build/gdev_api.o
CC [M] /homes/uriv/gdev/mod/build/gdev_device.o
CC [M] /homes/uriv/gdev/mod/build/gdev_sched.o
CC [M] /homes/uriv/gdev/mod/build/gdev_nvidia.o
CC [M] /homes/uriv/gdev/mod/build/gdev_nvidia_compute.o
CC [M] /homes/uriv/gdev/mod/build/gdev_nvidia_mem.o
CC [M] /homes/uriv/gdev/mod/build/gdev_nvidia_shm.o
CC [M] /homes/uriv/gdev/mod/build/gdev_nvidia_nvc0.o
LD [M] /homes/uriv/gdev/mod/build/gdev.o
Building modules, stage 2.
MODPOST 1 modules
WARNING: "gdev_drv_getdrm" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_bo_unmap" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_chan_alloc" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_write" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_vspace_free" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_bo_alloc" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_getdevice" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_vspace_alloc" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_getparam" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_bo_map" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_read" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_setnotify" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_read32" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_bo_bind" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_getaddr" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_unsetnotify" [q/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_bo_unbind" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_bo_free" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_write32" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
WARNING: "gdev_drv_chan_free" [/homes/uriv/gdev/mod/build/gdev.ko] undefined!
CC /homes/uriv/gdev/mod/build/gdev.mod.o
LD [M] /homes/uriv/gdev/mod/build/gdev.ko
make[1]: Leaving directory/homes/uriv/linux-2.6'

Please help us to resolve this issue.
Thank you.

**Comment by itziksh** [comment]
I made a mistake in my post,
we blacklisted nouveau, not envytools.

**Comment by mharsch** [comment]
pscnv needs to be built with gdev_interface.h and gdev_interface.c from gdev/mod/pscnv.  I'll see if I can come up with some better instructions.

**Comment by mharsch** [comment]
Taking a step back, I don't believe I was ever able to get the current version of pscnv to work with gdev.  @shinpei0208 can correct me if I'm wrong, but if I remember correctly: recent pscnv changes have broken the interfaces that gdev uses.  When I went down this path a couple months ago, I was only able to get (patched) nouveau to work with gdev.  I'm sure gdev could be made to work with the current pscnv, but I think it would require some investigation by someone who understands the interfaces better than I do.

**Comment by mharsch** [comment]
Specifically, I remember tracking down pathscale/pscnv@fef3e4b8ead25c672ce3798484edafc0dd77de6e as a pscnv flagday that broke gdev.

**Comment by pathscale** [comment]
I'm supportive of getting pscnv + gdev working together.  Please file a bug report with more details how we can unbreak this.

**Comment by shinpei0208** [comment]
mharsch, Thanks for your follow-up!
- Shinpei

**Comment by itziksh** [comment]
@cbergstrom , we will happily file a bug report, What details do you need more then we have provided? should we file the bug report in a different thread?

thanks.

**Comment by itziksh** [comment]
@mharsch We tried resetting git to df8fa305ea44a3158a600b200ff1677a0db3f487, which precedes fef3e4b8ead25c672ce3798484edafc0dd77de6e . However 'make' failed for pscnv with-

[ 11%] Generating ../../pscnv/nvc0_copy.fuc.h
no variant nva3
make[2]: **\* [../pscnv/nvc0_copy.fuc.h] Error 1
make[1]: **\* [pscnv/CMakeFiles/pscnv_ko.dir/all] Error 2
make: **\* [all] Error 2

**Comment by mharsch** [comment]
Unfortunately, gdev on current-pscnv is a dead-end right now.  We need to fix the way gdev talks to pscnv.  Unless someone else beats me to it, I will follow up with an investigation when I get back from vacation (~10 days from now).  In the mean time, you can try the historical version of pscnv bundled with gdev (running on kernel 2.6.38), or try patching nouveau per the patch bundled in gdev (kernel 2.6.38) or if you're running v3.5: https://raw.github.com/kris7topher/gdev/new-nouveau-patch/mod/patches/gdev-nouveau-3.5.patch

**Comment by wkbjerry** [comment]
I also met the `no variant nva3' error when I used the latest envytools in the compilation of gdev/mod/pscnv. Using an earlier version (I do not have the exact commit number) avoids the problem, but this is clear of a possible bug in the latest envytools/envyas.

As mharsch said, the original problem posted is due to pscnv compiled without gdev patch. I think it's better to use the pscnv driver within gdev (gdev/mod/pscnv) when you want to use gdev+pscnv for this moment. Hopefully gdev will catch up with nouveau/pscnv soon.

**Comment by mharsch** [comment]
see pathscale/pscnv@41b7e8e1743096e661cfb82224da7fdca5ada76c for the changes needed to pscnv to build with latest envytools.

**Comment by uriv** [comment]
Hi, I'm trying to get these things work together with @itziksh .

@wkbjerry  could you please check the safe envytools commit number? I'll try to find it too...

@mharsch tried with the pscnv commit you mention. The build is OK, but gdev fails with the same errors.

Thanks to both

**Comment by mharsch** [comment]
The latest envytools are fine, provided that the changes mentioned above are made.

The missing symbol messages are a red herring.  Most likely, the gdev_interface object file isn't getting linked properly using pscnv's cmake build infrastructure.  You should be able to get around this by building pscnv using the regular Makefile in the pscnv src dir.  That said, I stand by my claim that you're currently on a dead-end for the reasons previously mentioned.

**Comment by uriv** [comment]
The gdev pscnv isn't building either. says:

no variant nva3
no variant nva3
no variant nva3
nv98_crypt.fuc:241.46-242.1: syntax error

**Comment by wkbjerry** [comment]
What I meant was the gdev pscnv too. I just checked that the envytools
version which I used that works for the gdev pscnv is
https://0x04.net/indefero/index.php/p/envytools/source/tree/fc15385d454a8c032567c01ac58ead1b076de84f/

It is actually a very old one, commited in april.

**Comment by shinpei0208** [comment]
uriv, that error was due to a mismatching version of envytools. As wkbjerry suggests, please downgrade the envytools version accordingly.

mharsch, thanks for the v3.5 patch! I added it to gdev/mod/patches.
Please feel free to add the patch you make to the gdev repository henceforth :)

**Comment by uriv** [comment]
I downgraded to the suggested envytools version and used the pscnv provided with gdev. pscnv compiled and loaded ok. The gdev module and library steps went smoothly too. 

CUDA driver API worked too, but kcuda hit a compilation error at 'make'. 
The following changes made it compile:
in Makefile:
   EXTRA_CFLAGS = -I $(GDEVINC) -I $(DRMINC) -I /usr/include -I /usr/include/x86_64-linux-gnu
in gdev_cuda.c:
    in cubin_func_skip - changed 'printf' to GDEV_PRINT

Then, I installed CUDA toolkit (4.2), for nvcc, and tried running a test (user/madd). Now there is a linkage problem. 
I get:
nvcc -arch sm_20 -cubin -o madd_gpu.cubin madd_gpu.cu
gcc -o user_test -L /usr/local/gdev/lib64 -I /usr/local/gdev/include -lcuda -lgdev main.c madd.c
/tmp/ccDb6Bsa.o: In function `cuda_test_madd':
madd.c:(.text+0x21a): undefined reference to`cuInit'
madd.c:(.text+0x262): undefined reference to `cuDeviceGet'
...
madd.c:(.text+0xbe8): undefined reference to`cuCtxDestroy'
collect2: ld returned 1 exit status
make: **\* [all] Error 1

The cuda toolkit doesn't generate libcuda.so (the cuda driver does that and it is not installed), so the only library for -lcuda is in /usr/local/gdev/lib64/libcuda.so. I checked that the Makefile has -L /usr/local/gdev/lib64 in the compiler  flags and that cuInit is a symbol in libcuda.so using 'nm' command.

What could be the problem?

Thanks.

**Comment by wkbjerry** [comment]
Hi, uriv. You should use the kernel driver api test program, rather than
user-space one, when you use kcuda.

cd $(TOPDIR)/test/cuda/kernel/memcpy
make
sudo insmod ./kernel_test.ko size=10000 # copy 0x10000 size

I will confirm with your kcuda compilation errors, and post to you later.

Kaibo

**Comment by wkbjerry** [comment]
Hi, uriv. The error.h and printf bugs have been fixed. Thanks!

**Comment by uriv** [comment]
Hi @wkbjerry . Ok, so for running the kernel test I need to load kcuda and for the user test I need to not load kcuda, right? Yet, when I install the user API only, I get the same linkage problems as before. How do I run the user test?

The kernel test worked. I've noticed messages in the output that are probably a bug:
[10186.959443] DtoH: %f
[10186.959443] DataRead: %f
[10186.959444] Close: %f
[10186.959445] Total: %f
[10186.959445] Test passed

And, if we're into fixing  -- in the README file in the kcuda directions 'cd build' is missing :)

Thanks again

**Comment by wkbjerry** [comment]
Did you set the environment variables as mentioned in the README before runnning the user test?
export LD_LIBRARY_PATH="/usr/local/gdev/lib64:$LD_LIBRARY_PATH"
export PATH="/usr/local/gdev/bin:$PATH"

I'll fix `cd build' in README. Thanks :)

**Comment by uriv** [comment]
I did. gcc finds the library file. I tried to rename it and then it said that the library is missing.

**Comment by wkbjerry** [comment]
Did you compile and install user-space cuda driver api?

cd $(TOPDIR)/gdev/cuda
mkdir build
cd build
../configure
make
sudo make install

If that was done, then the other possible cause I can think of is the order of library search paths in LD_LIBRARY_PATH. /usr/local/gdev/lib64 should precede nvidia's if the latter exists

**Comment by uriv** [comment]
I went a small step further:
changed the order of arguments in the gcc statements in Makefile:
        #gcc -o $(TARGET) $(CFLAGS) $(LIBS) main.c madd.c
-->    gcc -o $(TARGET) $(CFLAGS) main.c madd.c $(LIBS)
apparently, this makes a difference in ubuntu http://stackoverflow.com/questions/7826448/linking-libraries-with-gcc-order-of-arguments
It compiles, now I get:
block = (16, 16)
grid = (16, 16)
cuCtxCreate failed: res = 999
Test failed

**Comment by shinpei0208** [comment]
Sorry I'm lost. Can you summarize your environment?
What GPU? Nouveau/pscnv? Linux kernel version? etc.

**Comment by uriv** [comment]
Environment:
Ubuntu 12.04.1 with Linux 3.3.0 x86_64
pscnv - old version suggested by @wkbjerry 
CUDA toolkit 4.2
two GTX 480 gpus
relevant loaded modules - gdev, pscnv, drm, drm_kms_helper, i2c_algo_bit, video

dmesg info:

there are messages such as:
[  240.263114] INFO: task gschedc0:1269 blocked for more than 120 seconds.
[  240.263274] INFO: task gschedm0:1270 blocked for more than 120 seconds.
...
at 240 and 360.

**Comment by shinpei0208** [comment]
uriv, What would happen if you switch to nouveau from pscnv?
That is, take Step 3.2 instead of 3.1.

**Comment by uriv** [comment]
It works with nouveau :) I didn't have to patch, just followed 3.2.

Do you want to get to the bottom of why it doesn't work with pscnv?

**Comment by wkbjerry** [comment]
When you used pscnv, did you check whether nouveau had been unloaded before loading pscnv? If pscnv were loaded when nouveau was functioning, the later cuda calls would fail.

**Comment by itziksh** [comment]
nouveau was blacklisted well in advance.

**Comment by wkbjerry** [comment]
BTW, the dmesg message:
[ 240.263114] INFO: task gschedc0:1269 blocked for more than 120 seconds.
[ 240.263274] INFO: task gschedm0:1270 blocked for more than 120 seconds.

does not matter. It is due to some GDev scheduling kernel threads idle for too long time. Just ignore it or use
echo 0 > /proc/sys/kernel/hung_task_timeout_secs
to depress it
